torchrl/collectors/collectors.py

```diff
@@ -820,69 +940,98 @@ def rollout(self) -> TensorDictBase:
+        self._final_rollout.fill_(("collector", "traj_ids"), -1)
         tensordicts = []
         with set_exploration_type(self.exploration_type):
             for t in range(self.frames_per_batch):
                 if (
                     self.init_random_frames is not None
                     and self._frames < self.init_random_frames
                 ):
-                    self.env.rand_action(self._tensordict)
+                    self.env.rand_action(self._shuttle)
                 else:
-                    self.policy(self._tensordict)
-                tensordict, tensordict_ = self.env.step_and_maybe_reset(
-                    self._tensordict
-                )
-                self._tensordict = tensordict_.set(
-                    "collector", tensordict.get("collector").clone(False)
-                )
-                tensordicts.append(
-                    tensordict.to(self.storing_device, non_blocking=True)
+                    if self._cast_to_policy_device:
+                        if self.policy_device is not None:
+                            policy_input = self._shuttle.to(
+                                self.policy_device, non_blocking=True
+                            )
+                        elif self.policy_device is None:
+                            # we know the tensordict has a device otherwise we would not be here
+                            policy_input = self._shuttle.copy().clear_device_()
+                    else:
+                        policy_input = self._shuttle
+                    # we still do the assignment for security
+                    policy_output = self.policy(policy_input)
+                    # update is a no-op if identities match, so this is safe and efficient in all cases
+                    self._shuttle.update(policy_output, inplace=True)
+
+                if self._cast_to_policy_device:
+                    if self.env_device is not None:
+                        env_input = self._shuttle.to(self.env_device, non_blocking=True)
+                    elif self.env_device is None:
+                        # we know the tensordict has a device otherwise we would not be here
+                        env_input = self._shuttle.copy().clear_device_()
+                else:
+                    env_input = self._shuttle
+
+                env_output, env_next_output = self.env.step_and_maybe_reset(env_input)
+                self._shuttle.update(env_output, inplace=True)
+                if self.storing_device is not None:
+                    tensordicts.append(
+                        self._shuttle.clone().to(self.storing_device, non_blocking=True)
+                    )
+                else:
+                    tensordicts.append(self._shuttle.clone())
+
+                self._shuttle.update(
+                    env_next_output.set(
+                        "collector", env_output.get("collector").copy()
+                    ),
+                    inplace=True,
```

Comment:
So this piece of code is pretty annoying but I don't see how to do otherwise.

Here's the pitch:

We want to support outputs that are on more than one device. Imagine your env is on cpu and your policy on cuda. You build your collector with
```python
collector = SyncCollector(..., policy_device="cuda", env_device="cpu")
```
You will have the `"action"`s on cuda and the rest on cpu.
Obviously the tensordict you will get in output won't have a device.

The way I make this happen is that I build a `_shuttle` that has preallocated tensors on the right devices and I update it inplace after each call to policy or env.step.
I do that because before you call the policy you need to cast your data to cuda using `shuttle.to(policy_device)` and conversely when you call env.step you need to cast it to cpu using `shuttle.to(env_device)` (note that in this case we could just cast the input data since the env tells us what input keys it needs -- which isn't always the case with the policy).

Because we need to cast stuff back and forth from devices, we need a copy of `shuttle` that has the proper devices allocated.

This also means that we need to clone the shuttle to save it in the stack of outputs (since modifications of shuttle must happen in-place). 


The problems are:
1. cloning induces an overhead that we'd like to avoid
2. modifying tensors in-place will make it hard to compile the collection loop when the time comes.

One way out would be to eagerly switch to a more efficient behaviour whenever policy_device and env_device match (no casting to device, no updating inplace). I think we'll still need the cloning but maybe not.

cc @matteobettini @skandermoalla

Reply:
I might need to look more in the whole shuttle thing, but a first point that i think is important is:

In this more common case, which is that policy and env are on the same device, we need to avoid any cloning and casting at all costs.

I thought that this is what currently happens in collectors, so for me a hard requirement is that these PR changes do not slowdown any already existing use-case.

We can then discuss the efficiency of the new use-cases but I just want to see if we are at least aligned on this point.

Reply:
Alright! I got rid of all the clones and inplace modifications.
The throughput seems not to be affected by this PR.